libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/MapboxTripStarter.kt

```diff
@@ -0,0 +1,214 @@
+package com.mapbox.navigation.core.trip
+
+import android.annotation.SuppressLint
+import androidx.annotation.VisibleForTesting
+import com.mapbox.android.core.permissions.PermissionsManager
+import com.mapbox.navigation.base.ExperimentalPreviewMapboxNavigationAPI
+import com.mapbox.navigation.core.MapboxNavigation
+import com.mapbox.navigation.core.lifecycle.MapboxNavigationApp
+import com.mapbox.navigation.core.lifecycle.MapboxNavigationObserver
+import com.mapbox.navigation.core.replay.route.ReplayRouteSession
+import com.mapbox.navigation.core.replay.route.ReplayRouteSessionOptions
+import com.mapbox.navigation.core.trip.MapboxTripStarterExtra.MAPBOX_TRIP_STARTER_FOLLOW_DEVICE
+import com.mapbox.navigation.core.trip.MapboxTripStarterExtra.MAPBOX_TRIP_STARTER_REPLAY_ROUTE
+import com.mapbox.navigation.core.trip.session.TripSessionState
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.combine
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+import kotlin.jvm.Throws
+
+/**
+ * This makes it simpler to start a MapboxNavigation trip session. It can be used to enable replay
+ * or to follow the device location. The [MapboxTripStarter] is not able to observe when location
+ * permissions change, so you must notify it of changes. It will check location permissions upon
+ * attach. The location permissions are not required for replay.
+ *
+ * There should be one instance of this class at a time. For example, an app Activity and car
+ * Session will need to use the same instance. That will be done automatically if you use
+ * [getRegisteredInstance].
+ */
+@ExperimentalPreviewMapboxNavigationAPI
+class MapboxTripStarter internal constructor() : MapboxNavigationObserver {
+
+    private val optionsFlow = MutableStateFlow(MapboxTripStarterOptions.Builder().build())
+    private val stateFlow = MutableStateFlow(MapboxTripStarterState())
+    private var replayRouteTripSession: ReplayRouteSession? = null
+    private var mapboxNavigation: MapboxNavigation? = null
```

Comment:
Why do we need to maintain class level `mapboxNavigation` field? The only use of it I see is inside the function `checkState()`, where this function gets access to context from `mapboxNavigation`. I am wondering if we can inject `mapboxNavigation` to this function
@kmadsen

Reply:
@abhishek1508 take another look cause I just refactored a bit. Supporting `checkState` was a pain in unit tests.

Now there is a `refreshLocationPermissions` which uses `mapboxNavigation` to access the `context`. I don't want to pass in the context, because it works without it.